src/main/java/com/ticketbooking/service/ReservationService.java

```diff
@@ -0,0 +1,323 @@
+            .orElseThrow(getNotFoundExceptionSupplier("screening.not.found", new String[] {String.valueOf(reservationDto.getScreeningId())}));
+
+        if (LocalDateTime.now().plusMinutes(15).isAfter(screening.getScreeningTime())) {
+            response.append(getMessage("booking.time.expired", null));
+        } else if (reservationDto.getTickets().isEmpty()) {
+            response.append(getMessage("one.place.reservation.rule", null));
+        } else if (!checkClientName(reservationDto)) {
+            response.append(getMessage("name.reservation.rule", null));
+        } else if (!checkClientSurname(reservationDto)) {
+            response.append(getMessage("surname.reservation.rule", null));
+        } else if (!checkSingleSeat(reservationDto, screening)) {
+            response.append(getMessage("new.place.reservation.rule", null));
+        } else {
+            Reservation reservation = saveReservation(reservationDto, screening);
+            BigDecimal ticketsPrice = saveTickets(reservationDto, screening, reservation);
+
+            sendReservationByMail(request, reservation.getId(), reservationDto.getMail());
+            
+            response.append(getMessage("confirm.reservation.line.one", null));
+            response.append(getMessage("confirm.reservation.line.two", new String[] {reservationDto.getName(), reservationDto.getSurname()}));
+            response.append(getMessage("confirm.reservation.line.three", new String[] {ticketsPrice.toString()}));
+            response.append(getMessage("confirm.reservation.line.four", null));
+        }
+
+        return response.toString();
+    }
+
+    private boolean checkClientSurname(ReservationDto reservationDto) {
+        if(reservationDto.getSurname() == null || reservationDto.getSurname().length() < 3) return false;
+
+        String words[] = reservationDto.getSurname().split("\\-");
+        StringBuilder capitalizeWord = new StringBuilder();
+        for(String w:words){
+            String first = w.substring(0,1);
+            String afterfirst = w.substring(1);
+            capitalizeWord.append(first.toUpperCase());
+            capitalizeWord.append(afterfirst);
+            capitalizeWord.append(" ");
+        }
+
+        reservationDto.setSurname(capitalizeWord.toString().trim().replace(" ", "-"));
+
+        return true;
+    }
+
+    private boolean checkClientName(ReservationDto reservationDto) {
+
+        if(reservationDto.getName() == null || reservationDto.getName().length() < 3) return false;
+        reservationDto.setName(reservationDto.getName().substring(0, 1).toUpperCase() + reservationDto.getName().substring(1));
+
+        return true;
+    }
+
+    private void sendReservationByMail(HttpServletRequest request, Long reservationId, String mail) {
+        try {
+            String url = "http://" +
+                    request.getServerName() + ":" + request.getServerPort() + request.getContextPath() +
```

Comment:
Czy ta cała magia zadziała, jeśli serwer będzie stał za np. reverse proxy? Jak to zrobić lepiej?

Reply:
Tak jak wcześniej napisałem eleganckim rozwiązaniem jest wykorzystanie mechanizu Event i Listener. Wszystkie potrzebne dane do wysłania maila są załączone do obiektu event, natomiast wysłanie maila odbyło by się w listenerze.

Reply:
Jak to zadziała z reverse proxy to dokładnie nie wiem. A w lepszym wykonaniu można użyć mechanizmu event/listener, i wtedy przenikanie warst będzie odizolowane.